clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/cache/client/EcsTargetGroupCacheClient.java

```diff
@@ -0,0 +1,134 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.ecs.cache.client;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spinnaker.cats.cache.Cache;
+import com.netflix.spinnaker.cats.cache.CacheData;
+import com.netflix.spinnaker.cats.cache.RelationshipCacheFilter;
+import com.netflix.spinnaker.clouddriver.aws.data.Keys;
+import com.netflix.spinnaker.clouddriver.ecs.model.loadbalancer.EcsTargetGroup;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.netflix.spinnaker.clouddriver.core.provider.agent.Namespace.LOAD_BALANCERS;
+import static com.netflix.spinnaker.clouddriver.core.provider.agent.Namespace.TARGET_GROUPS;
+
+@Component
+public class EcsTargetGroupCacheClient {
+
+  private final Cache cacheView;
+  private final ObjectMapper objectMapper;
+
+  public EcsTargetGroupCacheClient(Cache cacheView, ObjectMapper objectMapper) {
+    this.cacheView = cacheView;
+    this.objectMapper = objectMapper;
+  }
+
+  public List<EcsTargetGroup> findAll() {
+    String searchKey = Keys.getTargetGroupKey("*", "*", "*", "*") + "*";
+    Collection<String> targetGroupKeys = cacheView.filterIdentifiers(TARGET_GROUPS.getNs(), searchKey);
+
+    Set<Map<String, Object>> targetGroupAttributes = fetchLoadBalancerAttributes(targetGroupKeys);
+
+    List<EcsTargetGroup> targetGroups = convertToTargetGroup(targetGroupAttributes);
+
+    return targetGroups;
+  }
+
+  private EcsTargetGroup convertToTargetGroup(Map<String, Object> targetGroupAttributes) {
+    targetGroupAttributes.toString();
+    EcsTargetGroup ecsTargetGroup = objectMapper.convertValue(targetGroupAttributes, EcsTargetGroup.class);
+    return ecsTargetGroup;
+  }
+
+  private List<EcsTargetGroup> convertToTargetGroup(Collection<Map<String, Object>> targetGroupAttributes) {
+    List<EcsTargetGroup> ecsTargetGroups = new ArrayList<>();
+
+    for (Map<String, Object> attributes : targetGroupAttributes) {
+      ecsTargetGroups.add(convertToTargetGroup(attributes));
+    }
+
+    return ecsTargetGroups;
+  }
+
+
+  private Set<Map<String, Object>> fetchLoadBalancerAttributes(Collection<String> targetGroupKeys) {
+    Set<CacheData> targetGroups = fetchTargetGroups(targetGroupKeys);
+
+    Set<CacheData> targetGroupsWithAtLeastOneLoadBalancer = targetGroups
+      .stream()
+      .filter(targetGroup -> targetGroup.getRelationships().get("loadBalancers") != null
+        && targetGroup.getRelationships().get("loadBalancers").size() > 0)
+      .collect(Collectors.toSet());
+
+    Set<Map<String, Object>> targetGroupAttributes = targetGroupsWithAtLeastOneLoadBalancer.stream()
+      .map(lb -> lb.getAttributes())
+      .collect(Collectors.toSet());
```

Comment:
Should probably just combine these two streams.

```
// ...
.filter(EcsTargetGroupCacheClient::hasLoadBalancers)
.map(lb -> lb.getAttributes)
.collect(Collectors.toSet());
```

Bonus points for moving the filter logic into its own function so it's a little easier to read (purely my opinion here).

Reply:
Agreed, it does look better/easier to read.

Done.